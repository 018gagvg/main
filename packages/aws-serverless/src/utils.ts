import type { Scope } from '@sentry/types';
import { addExceptionMechanism } from '@sentry/utils';

/**
 * Marks an event as unhandled by adding a span processor to the passed scope.
 */
export function markEventUnhandled(scope: Scope): Scope {
  scope.addEventProcessor(event => {
    addExceptionMechanism(event, { handled: false });
    return event;
  });

  return scope;
}
