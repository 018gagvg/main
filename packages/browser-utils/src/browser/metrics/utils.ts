import type { SentrySpan } from '@sentry/core';
import { spanToJSON, startInactiveSpan, withActiveSpan } from '@sentry/core';
import type { Span, SpanTimeInput, StartSpanOptions } from '@sentry/types';

/**
 * Checks if a given value is a valid measurement value.
 */
export function isMeasurementValue(value: unknown): value is number {
  return typeof value === 'number' && isFinite(value);
}

/**
 * Helper function to start child on transactions. This function will make sure that the transaction will
 * use the start timestamp of the created child span if it is earlier than the transactions actual
 * start timestamp.
 *
 * Note: this will not be possible anymore in v8,
 * unless we do some special handling for browser here...
 */
export function startAndEndSpan(
  parentSpan: Span,
  startTimeInSeconds: number,
  endTime: SpanTimeInput,
  { ...ctx }: StartSpanOptions,
): Span | undefined {
  const parentStartTime = spanToJSON(parentSpan).start_timestamp;
  if (parentStartTime && parentStartTime > startTimeInSeconds) {
    // We can only do this for SentrySpans...
    if (typeof (parentSpan as Partial<SentrySpan>).updateStartTime === 'function') {
      (parentSpan as SentrySpan).updateStartTime(startTimeInSeconds);
    }
  }

  // The return value only exists for tests
  return withActiveSpan(parentSpan, () => {
    const span = startInactiveSpan({
      startTime: startTimeInSeconds,
      ...ctx,
    });

    if (span) {
      span.end(endTime);
    }

    return span;
  });
}
