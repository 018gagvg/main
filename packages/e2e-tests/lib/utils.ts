/* eslint-disable no-console */
import * as childProcess from 'child_process';

// https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-error-message
export function printCIErrorMessage(message: string): void {
  if (process.env.CI) {
    console.log(`::error::${message}`);
  } else {
    console.log(message);
  }
}

interface SpawnAsync {
  stdout: string;
  stderr: string;
  error?: Error;
  status: number | null;
}

interface SpawnOptions {
  timeout?: number;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  env?: any;
  cwd?: string;
}

export function spawnAsync(cmd: string, options?: SpawnOptions, input?: string): Promise<SpawnAsync> {
  const timeoutMs = options?.timeout || 60_000 * 5;

  return new Promise<SpawnAsync>(resolve => {
    const cp = childProcess.spawn(cmd, { shell: true, ...options });

    // Ensure we properly time out after max. 5 min per command
    let timeout: ReturnType<typeof setTimeout> | undefined = setTimeout(() => {
      console.log(`Command "${cmd}" timed out after 5 minutes.`);
      cp.kill();
      end(null, `ETDIMEDOUT: Process timed out after ${timeoutMs} ms.`);
    }, timeoutMs);

    const stderr: unknown[] = [];
    const stdout: string[] = [];
    let error: Error | undefined;

    function end(status: number | null, errorMessage?: string): void {
      // This means we already ended
      if (!timeout) {
        return;
      }

      if (!error && errorMessage) {
        error = new Error(errorMessage);
      }

      clearTimeout(timeout);
      timeout = undefined;
      resolve({
        stdout: stdout.join(''),
        stderr: stderr.join(''),
        error: error || (status !== 0 ? new Error(`Process exited with status ${status}`) : undefined),
        status,
      });
    }

    cp.stdout.on('data', data => {
      stdout.push(data ? (data as object).toString() : '');
    });

    cp.stderr.on('data', data => {
      stderr.push(data ? (data as object).toString() : '');
    });

    cp.on('error', e => {
      error = e;
    });

    cp.on('close', status => {
      end(status);
    });

    if (input) {
      cp.stdin.write(input);
      cp.stdin.end();
    }
  });
}

export function prefixObjectKeys(
  obj: Record<string, string | undefined>,
  prefix: string,
): Record<string, string | undefined> {
  return Object.keys(obj).reduce<Record<string, string | undefined>>((result, key) => {
    result[prefix + key] = obj[key];
    return result;
  }, {});
}
