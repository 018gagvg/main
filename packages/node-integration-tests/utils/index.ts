/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { Express } from 'express';
import * as http from 'http';
import nock from 'nock';
import { getPortPromise } from 'portfinder';

const keysToReplace = {
  timestamp: expect.any(Number),
  event_id: expect.any(String),
  filename: expect.any(String),
  version: expect.any(String),
  lineno: expect.any(Number),
};

const deepReplace = (obj: Record<string, any>, keyName: string, replacer: (from: any) => string): void => {
  for (const key in obj) {
    if (key === keyName) {
      obj[key] = replacer(obj[key]);
    } else if (Array.isArray(obj[key])) {
      (obj[key] as any[]).forEach(member => deepReplace(member, keyName, replacer));
    } else if (typeof obj[key] === 'object') {
      deepReplace(obj[key], keyName, replacer);
    }
  }
};

const updateForSnapshot = (obj: Record<string, any>): Record<string, any> => {
  Object.entries(keysToReplace).forEach(([key, value]) => {
    deepReplace(obj, key, () => value);
  });

  return obj;
};

const parseEnvelope = (body: string): Record<string, unknown> => {
  const [envelopeHeaderString, itemHeaderString, itemString] = body.split('\n');

  return {
    envelopeHeader: JSON.parse(envelopeHeaderString),
    itemHeader: JSON.parse(itemHeaderString),
    item: JSON.parse(itemString),
  };
};

const getEventRequest = async (url: string): Promise<Record<string, any>> => {
  return new Promise(resolve => {
    nock('https://dsn.ingest.sentry.io')
      .post('/api/1337/store/', body => {
        resolve(body);
        return true;
      })
      .reply(200);

    http.get(url);
  });
};

const getEnvelopeRequest = async (url: string): Promise<Record<string, any>> => {
  return new Promise(resolve => {
    nock(/ingest\.sentry\.io/)
      .post(/api\/1337\/envelope/, body => {
        resolve(body);
        return true;
      })
      .reply(200);

    http.get(url);
  });
};

async function runServer(testDir: string): Promise<string> {
  const port = await getPortPromise();
  const url = `http://localhost:${port}/test`;

  await new Promise(resolve => {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const app = require(`${testDir}/server`).default as Express;

    app.get('/test', async () => {
      require(`${testDir}/init`);
      require(`${testDir}/subject`);

      setTimeout(() => server.close(), 500);
    });

    const server = app.listen(port, () => {
      resolve();
    });
  });

  return url;
}

export { parseEnvelope, getEventRequest, getEnvelopeRequest, updateForSnapshot, runServer };
