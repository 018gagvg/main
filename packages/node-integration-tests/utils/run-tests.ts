/* eslint-disable no-console */
import childProcess from 'child_process';
import os from 'os';

const testPaths = childProcess.execSync('jest --listTests', { encoding: 'utf8' }).trim().split('\n');

let testsSucceeded = true;
const testAmount = testPaths.length;
const fails: string[] = [];

const threads = os.cpus().map(async (_, i) => {
  while (testPaths.length > 0) {
    const testPath = testPaths.pop();
    console.log(`(Worker ${i}) Running test "${testPath}"`);
    await new Promise(resolve => {
      const jestProcess = childProcess.spawn('jest', ['--runTestsByPath', testPath as string, '--forceExit']);

      let output = '';

      jestProcess.stdout.on('data', (data: Buffer) => {
        output = output + data.toString();
      });

      jestProcess.stderr.on('data', (data: Buffer) => {
        output = output + data.toString();
      });

      jestProcess.on('error', error => {
        console.log(`(Worker ${i}) Error in test "${testPath}"`, error);
        console.log(output);
        resolve();
      });

      jestProcess.on('exit', exitcode => {
        console.log(`(Worker ${i}) Finished test "${testPath}"`);
        console.log(output);
        if (exitcode !== 0) {
          fails.push(`FAILED: "${testPath}"`);
          testsSucceeded = false;
        }
        resolve();
      });
    });
  }
});

void Promise.all(threads).then(() => {
  console.log('-------------------');
  console.log(`Successfully ran ${testAmount} tests.`);
  if (!testsSucceeded) {
    console.log('Not all tests succeeded:\n');
    fails.forEach(fail => {
      console.log(`‚óè ${fail}`);
    });
    process.exit(1);
  } else {
    console.log('All tests succeeded.');
    process.exit(0);
  }
});
