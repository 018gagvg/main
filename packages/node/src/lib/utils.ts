import { mkdir, mkdirSync, statSync } from 'fs';
import { dirname, join, resolve } from 'path';

const _0777 = parseInt('0777', 8);

function mkdirAsync(path: string, mode: number): Promise<void> {
  // We cannot use util.promisify here because that was only introduced
  // in Node 8 and we need to support older Node versions.
  return new Promise((res, reject) => {
    mkdir(path, mode, err => (err ? reject(err) : res()));
  });
}

/**
 * Recursively creates the given path.
 *
 * @param path A relative or absolute path to create.
 * @returns A promise that resolves when the path has been created.
 */
export async function mkdirp(path: string): Promise<void> {
  // tslint:disable-next-line:no-bitwise
  const mode = _0777 & ~process.umask();
  const realPath = resolve(path);

  try {
    return mkdirAsync(realPath, mode);
  } catch (err) {
    if (err.code === 'ENOENT') {
      await mkdirp(dirname(realPath));
      return mkdirAsync(realPath, mode);
    } else {
      try {
        if (!statSync(realPath).isDirectory()) {
          throw err;
        }
      } catch (_) {
        throw err;
      }
    }
  }
}

/**
 * Synchronous version of {@link mkdirp}.
 *
 * @param path A relative or absolute path to create.
 */
export function mkdirpSync(path: string): void {
  // tslint:disable-next-line:no-bitwise
  const mode = _0777 & ~process.umask();
  const realPath = resolve(path);

  try {
    mkdirSync(realPath, mode);
  } catch (err) {
    if (err.code === 'ENOENT') {
      mkdirpSync(dirname(realPath));
      mkdirSync(realPath, mode);
    } else {
      try {
        if (!statSync(realPath).isDirectory()) {
          throw err;
        }
      } catch (_) {
        throw err;
      }
    }
  }
}
