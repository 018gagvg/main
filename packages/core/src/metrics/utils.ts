import type { MeasurementUnit, Primitive } from '@sentry/types';
import { dropUndefinedKeys } from '@sentry/utils';
import type { MetricType } from './types';

/**
 * Generate bucket key from metric properties.
 */
export function getBucketKey(
  metricType: MetricType,
  name: string,
  unit: MeasurementUnit,
  tags: { [key: string]: string },
): string {
  const stringifiedTags = Object.entries(dropUndefinedKeys(tags)).sort() as unknown as string[];
  return [metricType, name, unit].concat(stringifiedTags).join('');
}
