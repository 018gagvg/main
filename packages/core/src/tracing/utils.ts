import type { Span, Transaction } from '@sentry/types';
import type { Scope } from '@sentry/types';
import { addNonEnumerableProperty } from '@sentry/utils';
import { getCurrentScope } from '../currentScopes';

import type { Hub } from '../hub';
import { getCurrentHub } from '../hub';

/**
 * Grabs active transaction off scope.
 *
 * @deprecated You should not rely on the transaction, but just use `startSpan()` APIs instead.
 */
export function getActiveTransaction<T extends Transaction>(maybeHub?: Hub): T | undefined {
  // eslint-disable-next-line deprecation/deprecation
  const hub = maybeHub || getCurrentHub();
  // eslint-disable-next-line deprecation/deprecation
  const scope = hub.getScope();
  // eslint-disable-next-line deprecation/deprecation
  return scope.getTransaction() as T | undefined;
}

// so it can be used in manual instrumentation without necessitating a hard dependency on @sentry/utils
export { stripUrlQueryAndFragment } from '@sentry/utils';

/**
 * Returns the currently active span.
 */
export function getActiveSpan(): Span | undefined {
  // eslint-disable-next-line deprecation/deprecation
  return getCurrentScope().getSpan();
}

const SCOPE_ON_START_SPAN_FIELD = '_sentryScope';
const ISOLATION_SCOPE_ON_START_SPAN_FIELD = '_sentryIsolationScope';

type SpanWithScopes = Span & {
  [SCOPE_ON_START_SPAN_FIELD]?: Scope;
  [ISOLATION_SCOPE_ON_START_SPAN_FIELD]?: Scope;
};

/** Store the scope & isolation scope for a span, which can the be used when it is finished. */
export function setCapturedScopesOnSpan(span: Span | undefined, scope: Scope, isolationScope: Scope): void {
  if (span) {
    addNonEnumerableProperty(span, ISOLATION_SCOPE_ON_START_SPAN_FIELD, isolationScope);
    addNonEnumerableProperty(span, SCOPE_ON_START_SPAN_FIELD, scope);
  }
}

/**
 * Grabs the scope and isolation scope off a span that were active when the span was started.
 */
export function getCapturedScopesOnSpan(span: Span): { scope?: Scope; isolationScope?: Scope } {
  return {
    scope: (span as SpanWithScopes)[SCOPE_ON_START_SPAN_FIELD],
    isolationScope: (span as SpanWithScopes)[ISOLATION_SCOPE_ON_START_SPAN_FIELD],
  };
}
