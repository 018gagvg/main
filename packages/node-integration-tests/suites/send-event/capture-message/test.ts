import { getEventRequest, runServer, updateForSnapshot } from '../../../utils';

test('should send captureMessage', async () => {
  const url = await runServer(__dirname);
  const requestBody = await getEventRequest(url);

  expect(updateForSnapshot(requestBody)).toMatchSnapshot();
});
