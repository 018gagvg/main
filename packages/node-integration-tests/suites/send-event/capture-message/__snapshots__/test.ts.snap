// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should send captureMessage 1`] = `
Object {
  "environment": "production",
  "event_id": Any<String>,
  "extra": Object {
    "normalizeDepth": 3,
  },
  "level": "info",
  "message": "Message",
  "platform": "node",
  "sdk": Object {
    "integrations": Array [
      "InboundFilters",
      "FunctionToString",
      "ContextLines",
      "Console",
      "Http",
      "OnUncaughtException",
      "OnUnhandledRejection",
      "LinkedErrors",
    ],
    "name": "sentry.javascript.node",
    "packages": Array [
      Object {
        "name": "npm:@sentry/node",
        "version": Any<String>,
      },
    ],
    "version": Any<String>,
  },
  "tags": Object {
    "skippedNormalization": true,
  },
  "timestamp": Any<Number>,
}
`;
